MaterialLib/MPL/Properties/PermeabilityMohrCoulombFailureIndexModel.cpp

```diff
@@ -0,0 +1,148 @@
+/**
+ * \file
+ * \copyright
+ * Copyright (c) 2012-2020, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ * Created on June 4, 2020, 10:13 AM
+ */
+
+#include "PermeabilityMohrCoulombFailureIndexModel.h"
+
+#include <cmath>
+
+#include "MaterialLib/MPL/Medium.h"
+#include "MathLib/KelvinVector.h"
+#include "MathLib/MathematicalConstant.h"
+#include "ParameterLib/CoordinateSystem.h"
+#include "ParameterLib/Parameter.h"
+
+namespace MaterialPropertyLib
+{
+template <int DisplacementDim>
+PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::
+    PermeabilityMohrCoulombFailureIndexModel(
+        std::string name, ParameterLib::Parameter<double> const& k0,
+        double const kr, double const b, double const c, double const phi,
+        ParameterLib::CoordinateSystem const* const local_coordinate_system)
+    : k0_(k0),
+      kr_(kr),
+      b_(b),
+      c_(c),
+      phi_(MathLib::MathematicalConstant::pi * phi / 180.0),
+      local_coordinate_system_(local_coordinate_system)
+{
+    name_ = std::move(name);
+}
+
+template <int DisplacementDim>
+void PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::checkScale()
+    const
+{
+    if (!std::holds_alternative<Phase*>(scale_))
+    {
+        OGS_FATAL(
+            "The property 'PermeabilityMohrCoulombFailureIndexModel' is "
+            "implemented on the 'phase' scales only.");
+    }
+
+    auto const phase = std::get<Phase*>(scale_);
+    if (phase->name != "Solid")
+    {
+        OGS_FATAL(
+            "The property 'PermeabilityMohrCoulombFailureIndexModel' must be "
+            "given in the 'Solid' phase, not in '{:s}' phase.",
+            phase->name);
+    }
+}
+
+template <int DisplacementDim>
+PropertyDataType
+PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::value(
+    VariableArray const& variable_array,
+    ParameterLib::SpatialPosition const& pos, double const t,
+    double const /*dt*/) const
+{
+    constexpr int symmetric_tensor_size =
+        MathLib::KelvinVector::KelvinVectorDimensions<DisplacementDim>::value;
+    using SymmetricTensor = Eigen::Matrix<double, symmetric_tensor_size, 1>;
+
+    auto const& stress_vector = std::get<SymmetricTensor>(
+        variable_array[static_cast<int>(Variable::stress)]);
+    auto const& stress_tensor =
+        MathLib::KelvinVector::symmetricTensorToFullTensor(stress_vector);
+
+    Eigen::SelfAdjointEigenSolver<Eigen::Matrix<double, 3, 3>>
+        eigenvalue_solver(stress_tensor);
+
+    // Principle stress
+    auto const sigma = eigenvalue_solver.eigenvalues();
+
+    double const f =
+        std::fabs(0.5 * (sigma[0] - sigma[2]) /
+                  (c_ + 0.5 * (sigma[0] + sigma[2]) * std::tan(phi_)));
+
```

Comment:
This is not what we have in the paper ... there we have:

        0.5 * std::fabs(sigma[0] - sigma[2]) /
                  (c_ * std::(cos(phi_) + 0.5 * (sigma[0] + sigma[2]) * std::sin(phi_));

which is the actual MohrCoulomb formulation (not maximum shear stress drives failure, but worst ratio of shear stress to normal stress).

The std::fabs around everything also does not seem to make sense in the hypothetical far tension area (beyond the hydrostatic tensile strength) .... this should be treated differently even if it will rarely arise ... 

By the way, if the stress state is at this point, the denominator will become zero and you get a NaN, so this should also be caught ...

Reply:
Thinking about this, I suggest three unit tests:
-- one stress point within the yield surface (you have that)
-- one stress point outside of the yield surface (you have that)
-- one stress point beyond the apex of the yield surface (to test this "special behaviour")

Reply:
> This is not what we have in the paper ... there we have:
> 
> ```
>     0.5 * std::fabs(sigma[0] - sigma[2]) /
>               (c_ * std::(cos(phi_) + 0.5 * (sigma[0] + sigma[2]) * std::sin(phi_));
> ```
> 
> which is the actual MohrCoulomb formulation (not maximum shear stress drives failure, but worst ratio of shear stress to normal stress).
> 
> The std::fabs around everything also does not seem to make sense in the hypothetical far tension area (beyond the hydrostatic tensile strength) .... this should be treated differently even if it will rarely arise ...
> 
> By the way, if the stress state is at this point, the denominator will become zero and you get a NaN, so this should also be caught ...

Thanks. The formula will be changed to the  actual MohrCoulomb one. The test for the stress point beyond the apex  will be added.

Reply:
> The std::fabs around everything also does not seem to make sense in the hypothetical far tension area (beyond the hydrostatic tensile strength) .... this should be treated differently even if it will rarely arise ...

Removed std:fabs

Reply:
> -- one stress point beyond the apex of the yield surface (to test this "special behaviour")

Added.